v2/protocol/http/protocol_retry.go

```diff
@@ -0,0 +1,123 @@
+package http
+
+import (
+	"context"
+	"errors"
+	"go.uber.org/zap"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+func (p *Protocol) do(ctx context.Context, req *http.Request) (binding.Message, error) {
+	params := cecontext.RetriesFrom(ctx)
+
+	switch params.Strategy {
+	case cecontext.BackoffStrategyConstant, cecontext.BackoffStrategyLinear, cecontext.BackoffStrategyExponential:
+		return p.doWithRetry(ctx, params)(req)
+	case cecontext.BackoffStrategyNone:
+		fallthrough
+	default:
+		return p.doOnce(req)
+	}
+}
+
+func (p *Protocol) doOnce(req *http.Request) (binding.Message, protocol.Result) {
+	resp, err := p.Client.Do(req)
+	if err != nil {
+		return nil, protocol.NewReceipt(false, "%w", err)
+	}
+
+	var result protocol.Result
+	if resp.StatusCode/100 == 2 {
+		result = protocol.ResultACK
+	} else {
+		result = protocol.ResultNACK
+	}
+
+	return NewMessage(resp.Header, resp.Body), NewResult(resp.StatusCode, "%w", result)
+}
+
+func (p *Protocol) doWithRetry(ctx context.Context, params *cecontext.RetryParams) func(*http.Request) (binding.Message, error) {
+	return func(req *http.Request) (binding.Message, error) {
+		params := params
+		then := time.Now()
+		retry := 0
+		results := make([]protocol.Result, 0)
+
+		for {
+			msg, result := p.doOnce(req)
+
+			// Fast track common case.
+			if protocol.IsACK(result) {
+				return msg, NewRetriesResult(result, retry, then, results)
+			}
+
+			// Try again?
+			//
+			// Make sure the error was something we should retry.
+
+			//{
+			//	var sErr *syscall.Errno
+			//	if errors.As(result, &sErr) {
+			//		// system level error, retry.
+			//		goto DO_BACKOFF
+			//	}
+			//}
+
+			{
+				var uErr *url.Error
+				if errors.As(result, &uErr) {
+					// Only retry when the error is a timeout
+					//if uErr.Timeout() || uErr.Temporary() {
```

Comment:
This seems reasonable to me. eg: If the hostname is invalid, there's no point on retrying

Reply:
it might change. There is no reason not to try. Like, take k8s, if the service is not up but will be soon... keep trying.

Reply:
We should not get to this point with an invalid hostname because we use url.URL parsed objects

Reply:
if the service is not up i assume that `Error.Temporary() == true` right? like unable to connect and stuff like that is temporary

Reply:
no, if the service is not up it is fatal. that is only temporary in k8s :D